PyDynamic/uncertainty/propagate_filter.py

```diff
@@ -167,99 +168,274 @@ def FIRuncFilter(y, sigma_noise, theta, Utheta=None, shift=0, blow=None, kind="c
+            unc[m] = XL.T.dot(Utheta.dot(XL))     # apply formula from paper
+    
     ux = np.sqrt(np.abs(UncCov + unc))
     ux = np.roll(ux, -int(shift))  # correct for delay
 
     return x, ux.flatten()  # flatten in case that we still have 2D array
 
 
-def IIRuncFilter(x, noise, b, a, Uab):
+def IIRuncFilter(x, Ux, b, a, Uab, state = None, kind="diag"):
     """
     Uncertainty propagation for the signal x and the uncertain IIR filter (b,a)
 
     Parameters
     ----------
-    x: np.ndarray
-        filter input signal
-    noise: float
-        signal noise standard deviation
-    b: np.ndarray
-        filter numerator coefficients
-    a: np.ndarray
-        filter denominator coefficients
-    Uab: np.ndarray
-        covariance matrix for (a[1:],b)
+        x: np.ndarray
+            filter input signal
+        Ux: float or np.ndarray
+            float:    standard deviation of white noise in y
+            1D-array: interpretation depends on kind
+        b: np.ndarray
+            filter numerator coefficients
+        a: np.ndarray
+            filter denominator coefficients
+        Uab: np.ndarray
+            covariance matrix for (a[1:],b)
+        kind: string
+            defines the interpretation of Ux, if Ux is a 1D-array
+            "diag": point-wise standard uncertainties of non-stationary white noise (default)
+            "corr": single sided autocovariance of stationary (colored/corrlated) noise
+        state: dict
+            An internal state (z, dz, P, cache) to start from - e.g. from a previous run of IIRuncFilter.
+            - If not given, (z, dz, P) are calculated such that the signal was constant before the given range
+            - If given, the input parameters (b, a, Uab) are ignored to avoid repetitive rebuild of the
+              internal system description (instead, the cache is used). However a valid new state (i.e.
+              with new b, a, Uab) can always be generated by using :py:func:`get_initial_state()`.
 
     Returns
     -------
-    y: np.ndarray
-        filter output signal
-    Uy: np.ndarray
-        uncertainty associated with y
+        y: np.ndarray
+            filter output signal
+        Uy: np.ndarray
+            uncertainty associated with y
+        state: dict
+            dictionary of updated internal state
+
+    Note
+    ----
+        In case of `a == [1.0]` (an FIR filter), the results of :py:func:`IIRuncFilter` and :py:func:`FIRuncFilter` will differ!
+        
+        This is because IIRuncFilter propagates uncertainty according to the 
+        (first-order Taylor series of the) GUM, wheras FIRuncFilter takes full 
+        variance information into account (which leads to an additonal term). 
+        This is documented in the description of formula (33) of [Elster2008]_ . 
+
+        The difference can be visualized by running :ref:`examples/Digital filtering/validate_FIR_IIR_MC.py`
```

Comment:
The examples are actually not available as references as such. We could either insert a link to the GitHub repo or just mention the file name and location.

Reply:
:+1: